lib/FrontendTool/ReferenceDependencies.cpp

```diff
@@ -22,90 +22,156 @@
-    case DeclKind::Extension:
-    case DeclKind::InfixOperator:
-    case DeclKind::PrefixOperator:
-    case DeclKind::PostfixOperator:
-      return false;
-
-    default:
-      llvm_unreachable("everything else is a ValueDecl");
-    }
-  }
-
-  return VD->getFormalAccess() <= AccessLevel::FilePrivate;
-}
-
-static bool extendedTypeIsPrivate(TypeLoc inheritedType) {
-  auto type = inheritedType.getType();
-  if (!type)
-    return true;
-
-  if (!type->isExistentialType()) {
-    // Be conservative. We don't know how to deal with other extended types.
-    return false;
-  }
-
-  auto layout = type->getExistentialLayout();
-  assert(!layout.superclass && "Should not have a subclass existential "
-         "in the inheritance clause of an extension");
-  for (auto protoTy : layout.getProtocols()) {
-    if (!declIsPrivate(protoTy->getDecl()))
-      return false;
-  }
-
-  return true;
-}
+  /// Emit all provided declartions.
+  void emit() const;
+
+  CollectedDeclarations emitTopLevelNames() const;
+  void emitNominalTypes(const llvm::MapVector<const NominalTypeDecl *, bool>
+                            &extendedNominals) const;
+  void emitMembers(const CollectedDeclarations &cpd) const;
+  void emitDynamicLookupMembers() const;
+
+  void emitTopLevelDecl(const Decl *D, CollectedDeclarations &cpd) const;
+  void emitExtensionDecl(const ExtensionDecl *D,
+                         CollectedDeclarations &cpd) const;
+  void emitNominalTypeDecl(const NominalTypeDecl *NTD,
+                           CollectedDeclarations &cpd) const;
+  void emitValueDecl(const ValueDecl *VD) const;
+
+  static bool extendedTypeIsPrivate(TypeLoc inheritedType);
+  static bool declIsPrivate(const Decl *member);
+};
+
+/// Emit the depended-upon declartions.
+class DependsEmitter {
+  /// The file that dependes upon the declarations.
+  const SourceFile *const SF;
+  /// The dependencies collected by the compiler.
+  const DependencyTracker &depTracker;
+
+  llvm::raw_ostream &out;
+
+  DependsEmitter(const SourceFile *SF, const DependencyTracker &depTracker,
+                 llvm::raw_ostream &out)
+      : SF(SF), depTracker(depTracker), out(out) {}
+
+public:
+  /// A NominalTypeDecl, its DeclBaseName, and whether it is externally-visible.
+  using MemberTableEntryTy = std::pair<ReferencedNameTracker::MemberPair, bool>;
+
+  /// Emit the dependencies
+  ///
+  /// \param SF SourceFile containing the dependent code
+  /// \param depTracker Contains the dependencies found during compilation
+  /// \param out Where the dependencies are emitted
+  static void emit(const SourceFile *SF, const DependencyTracker &depTracker,
+                   llvm::raw_ostream &out);
+
+private:
+  /// Emit all the dependencies.
+  void emit() const;
+
+  void emitTopLevelNames(const ReferencedNameTracker *const tracker) const;
+  void emitMember(const ArrayRef<MemberTableEntryTy> sortedMembers) const;
+  void emitNominal(const ArrayRef<MemberTableEntryTy> sortedMembers) const;
```

Comment:
Shouldn't these be "Members" and "NominalTypes" for consistency with the Provides emitter?